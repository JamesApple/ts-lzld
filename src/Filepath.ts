import path from 'path';
interface Data {
  absolute: string;
  type: 'file' | 'dir';
}

export class AbsolutePath {
  static fromAbsolute(absolute: string, type: 'file' | 'dir'): AbsolutePath {
    return new AbsolutePath({ absolute, type });
  }

  constructor(private data: Data) {}

  get absolute(): string {
    return this.data.absolute;
  }

  get completeFilename(): string {
    return path.basename(this.absolute);
  }

  getImportablePathTo(importTarget: AbsolutePath): string {
    return this.relativePathTo(importTarget.withoutImportableExtension);
  }

  get type(): 'file' | 'dir' {
    return this.data.type;
  }

  get dirname(): string {
    if (this.data.type === 'file') {
      return path.dirname(this.absolute);
    } else {
      return this.absolute;
    }
  }

  relativePathTo(target: AbsolutePath): string {
    const relative = path.relative(this.dirname, target.absolute);
    if (relative.startsWith('..')) {
      return relative;
    } else {
      return './' + relative;
    }
  }

  addRelative(relativePath: string, type: 'file' | 'dir'): AbsolutePath {
    return AbsolutePath.fromAbsolute(
      path.resolve(this.dirname, relativePath),
      type
    );
  }

  get withoutImportableExtension(): AbsolutePath {
    if (this.data.type === 'dir') {
      return this;
    }
    const extensions = ['.ts', '.tsx', '.js', '.jsx'];
    const ext = extensions.find(ext => this.absolute.endsWith(ext));
    if (ext) {
      return AbsolutePath.fromAbsolute(
        this.absolute.slice(0, -ext.length),
        'file'
      );
    }
    return AbsolutePath.fromAbsolute(this.absolute, 'file');
  }
}
