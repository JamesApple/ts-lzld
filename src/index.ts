export { Entrypoint } from './Entrypoint';
export {
  MetadataLike,
  GeneratedFileData,
  MatchedPath,
  ConstructorLike,
  Entry,
} from './types';
export { EntrypointOptions } from './options';
