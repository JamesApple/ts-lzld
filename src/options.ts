import path from 'path';
import { ConstructorLike, MatchedPath, MetadataLike } from './types';

export interface EntrypointOptions<
  T extends ConstructorLike = ConstructorLike,
  M extends MetadataLike = MetadataLike
> {
  /**
   * Which typescript files export an entrypoint.
   * @example
   * ```typescript
   * /([a-zA-Z0-9]+).handler.ts$/
   * ```
   */
  match: RegExp;
  /**
   * Metadata about the entrypoint that can be used to locate the correct
   * entrypoint.
   * @example
   * ```typescript
   * (handler, { result: [, name] }) => ({
   *   path: handler.http.path,
   *   method: handler.http.method,
   *   name,
   * })
   * ```
   */
  getMetadata?: (target: T, match: MatchedPath) => M;
  /**
   * Allows changing the root where entrypoints will be loaded from.
   * @default '.'
   */
  searchPath?: string;
  metadataFilepath: string;
  /**
   * This should be the nodejs global __filename so that require statements can
   * be correctly resolved and generated.
   *
   * Using anything other than `__filename` will result in incorrect behaviour.
   */
  __filename: string;

  /**
   * # Whether we should generate a metadata file for this entrypoint.
   *
   * This imports all the entrypoints and generates a metadata file using their
   * class names and the metadata returned by `getMetadata`.
   *
   * @default
   * ```typescript
   *   process.env.NODE_ENV === 'development'
   * ```
   */
  generate?: boolean;
}

export interface CompleteEntrypointOptions {
  match: RegExp;
  getMetadata: (target: ConstructorLike, match: MatchedPath) => MetadataLike;

  sourceFilename: string;

  absoluteMetadataFilePath: string;
  localPathPrefix: string;
  absolutePathPrefix: string;

  generate: boolean;
}

export const completeOptions = (
  opts: EntrypointOptions<any, any>
): CompleteEntrypointOptions => {
  const dirname = path.dirname(opts.__filename);
  return {
    getMetadata: opts.getMetadata ?? (() => ({})),
    match: opts.match,
    sourceFilename: path.basename(opts.__filename),
    localPathPrefix: opts.searchPath ?? '.',
    absolutePathPrefix: path.resolve(dirname, opts.searchPath ?? ''),
    generate: opts.generate ?? process.env.NODE_ENV === 'development',
    absoluteMetadataFilePath: path.resolve(dirname, opts.metadataFilepath),
  };
};
