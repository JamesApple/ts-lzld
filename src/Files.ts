import fs from 'fs';
import { AbsolutePath } from './Filepath';

export interface Files {
  createFile(filepath: AbsolutePath, content: string): void;
  recurseDirectory(
    filepath: AbsolutePath,
    callback: (path: AbsolutePath) => unknown
  ): void;

  require(path: AbsolutePath): any;
}

export class NodeFiles implements Files {
  require(path: AbsolutePath): any {
    return require(path.absolute);
  }

  createFile(filepath: AbsolutePath, content: string): void {
    if (filepath.type !== 'file') {
      throw new Error(
        'Cannot create a file to a directory path ' + filepath.absolute
      );
    }

    if (!fs.existsSync(filepath.dirname)) {
      fs.mkdirSync(filepath.dirname, { recursive: true });
    }
    if (fs.existsSync(filepath.absolute)) {
      if (fs.readFileSync(filepath.absolute, 'utf8') === content) {
        return;
      }
    }

    fs.writeFileSync(filepath.absolute, content, {
      encoding: 'utf8',
    });
  }

  recurseDirectory(
    filepath: AbsolutePath,
    callback: (path: AbsolutePath) => unknown
  ): void {
    if (filepath.type !== 'dir') {
      throw new Error(
        'Cannot recurse a directory from a file path' + filepath.absolute
      );
    }

    fs.readdirSync(filepath.absolute).forEach(relativePath => {
      const { absolute } = filepath.addRelative(relativePath, 'file');

      const stat = fs.statSync(absolute);

      if (stat.isDirectory()) {
        this.recurseDirectory(
          AbsolutePath.fromAbsolute(absolute, 'dir'),
          callback
        );
      }

      if (stat.isFile()) {
        callback(AbsolutePath.fromAbsolute(absolute, 'file'));
      }
    });
  }
}

interface Data {
  // eslint-disable-next-line @typescript-eslint/ban-types
  modules: { [filepath: string]: Object };
  createdFiles: { [filepath: string]: string };
  recurseResult: { [filepath: string]: AbsolutePath[] };
}

export const makeInMemoryFiles = (args: Partial<Data>): [Files, Data] => {
  const data: Data = {
    modules: {},
    createdFiles: {},
    recurseResult: {},
    ...args,
  };

  const files: Files = {
    createFile(filepath: AbsolutePath, content: string): void {
      data.createdFiles[filepath.absolute] = content;
    },
    recurseDirectory(
      filepath: AbsolutePath,
      callback: (path: AbsolutePath) => unknown
    ): void {
      if (!data.recurseResult[filepath.absolute]) {
        throw new Error(`No recurse result for ${filepath.absolute}`);
      }

      data.recurseResult[filepath.absolute].forEach(callback);
    },
    require(path: AbsolutePath): any {
      if (!data.modules[path.absolute]) {
        throw new Error(`No module for ${path.absolute}`);
      }
      return data.modules[path.absolute];
    },
  };

  return [files, data];
};
