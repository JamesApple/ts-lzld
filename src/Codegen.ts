import { AbsolutePath } from './Filepath';
import { CompleteEntrypointOptions } from './options';
import { Runtime } from './Runtime';
import { ConstructorLike, MetadataLike } from './types';

export class Codegen<T extends ConstructorLike, M extends MetadataLike> {
  constructor(
    private filepath: AbsolutePath,
    private opts: CompleteEntrypointOptions,
    private runtime: Runtime<T, M>
  ) {}

  template: TemplaterFunction<T, M> = (fragments, ...funcs) => {
    if (!this.shouldGenerate) {
      return;
    }

    const makeImportFn = (name: string, path: string) => (importAs: string) => {
      if (name === 'default') {
        return `import ${importAs} from '${path}';`;
      }
      if (importAs === name) {
        return `import { ${name} } from '${path}';`;
      }
      return `import { ${name} as ${importAs} } from '${path}';`;
    };

    const exportFn = (name: string, path: string) => (importAs: string) => {
      if (importAs === name) {
        return `export { ${name} } from '${path}';`;
      }
      return `export { ${name} as ${importAs} } from '${path}';`;
    };

    const entries: EntrypointInfo<T, M>[] = this.runtime.entries.map(e => {
      const absoluteEntryPath = this.runtime.opts.searchPath.addRelative(
        e.path,
        'file'
      );
      const importablePath =
        this.filepath.getImportablePathTo(absoluteEntryPath);

      return {
        entry: this.runtime.load(e).entry,
        importablePath,
        exportedAs: e.name,
        meta: e.meta,
        importStatement: makeImportFn(e.name, importablePath),
        exportStatement: exportFn(e.name, importablePath),
      };
    });

    const data = fragments.reduce(
      (current, fragment, index) =>
        `${current}${this.resolveTemplateString(
          funcs[index - 1],
          entries
        )}${fragment}`,
      ''
    );

    this.opts.files.createFile(this.filepath, data);
  };

  private resolveTemplateString(
    func: GenerateFunction<T, M> | undefined,
    entries: EntrypointInfo<T, M>[]
  ) {
    if (func == null) {
      return '';
    }
    if (typeof func === 'string') {
      return func;
    }
    const result = func({ entries });
    if (Array.isArray(result)) {
      return result.join('\n');
    }
    return result;
  }

  private get shouldGenerate() {
    return this.runtime.opts.generate;
  }
}

interface EntrypointInfo<T extends ConstructorLike, M extends MetadataLike> {
  entry: T;
  meta: M;
  exportedAs: string;
  importablePath: string;
  importStatement: (as: string) => string;
  exportStatement: (as: string) => string;
}

interface GenerateArgs<T extends ConstructorLike, M extends MetadataLike> {
  entries: EntrypointInfo<T, M>[];
}

type GenerateFunction<T extends ConstructorLike, M extends MetadataLike> =
  | ((args: GenerateArgs<T, M>) => string | string[])
  | string;

export interface TemplaterFunction<
  T extends ConstructorLike,
  M extends MetadataLike
> {
  (
    fragments: TemplateStringsArray,
    ...generateFunctions: GenerateFunction<T, M>[]
  ): void;
}
