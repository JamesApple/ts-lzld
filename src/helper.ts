import { AbsolutePath } from './Filepath';
import { Files } from './Files';
import { ConstructorLike } from './types';

export interface ResolvedEntrypoint<T extends ConstructorLike> {
  name: string;
  entry: T;
}

export const resolveEntrypoint = <T extends ConstructorLike>(
  supertype: T,
  absolutePath: AbsolutePath,
  files: Files
): ResolvedEntrypoint<T> => {
  const resolved = files.require(absolutePath);
  const exports = Object.entries(resolved);

  const entrypoint = exports.find(([, exported]) => {
    if (typeof exported !== 'function') return false;
    return exported.prototype instanceof supertype;
  });

  if (!entrypoint) {
    throw new Error(
      `lzld expected a ${supertype.name} to be exported from ${absolutePath.absolute} but found none`
    );
  }

  return {
    name: entrypoint[0],
    entry: entrypoint[1] as T,
  };
};
