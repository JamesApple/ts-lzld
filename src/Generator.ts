import {
  ConstructorLike,
  Entry,
  GeneratedFileData,
  MatchedPath,
  MetadataLike,
} from './types';
import { CompleteEntrypointOptions } from './options';
import { resolveEntrypoint } from './helper';

export class Generator {
  constructor(
    private supertype: ConstructorLike,
    private opts: CompleteEntrypointOptions
  ) {}

  generate(): void {
    const { files } = this.opts;

    const entries = this.getEntries();

    const fileData: GeneratedFileData = {
      '//': `This file is auto-generated in '${this.opts.sourceFile.completeFilename}'`,
      lzld: 'generated',
      entries,
    };

    files.createFile(this.opts.metadataFile, JSON.stringify(fileData, null, 2));
  }

  private getEntries(): Entry<MetadataLike>[] {
    const { files, searchPath, getMetadata } = this.opts;
    const matches = this.getMatches();

    return matches.map(match => {
      const relative = searchPath.getImportablePathTo(match.path);
      const { name, entry } = resolveEntrypoint(
        this.supertype,
        match.path,
        files
      );
      return {
        name,
        path: relative,
        meta: getMetadata(entry, match),
      };
    });
  }

  private getMatches(): MatchedPath[] {
    const paths: MatchedPath[] = [];
    const { match: matcher, searchPath, files } = this.opts;

    files.recurseDirectory(searchPath, path => {
      if (path.type === 'file') {
        const relativePath = searchPath.relativePathTo(path);
        const result = matcher.exec(relativePath);

        if (result) {
          paths.push({ path, result });
        }
      }
    });

    return paths;
  }
}
