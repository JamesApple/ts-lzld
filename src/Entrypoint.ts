import { Codegen, TemplaterFunction } from './Codegen';
import { Generator } from './Generator';
import {
  CompleteEntrypointOptions,
  completeOptions,
  EntrypointOptions,
} from './options';
import { Runtime } from './Runtime';
import { ConstructorLike, LoadableMetadata, MetadataLike } from './types';

export class Entrypoint<T extends ConstructorLike, M extends MetadataLike> {
  static init: Init = (supertype, args) => {
    const options = completeOptions(args);

    if (options.generate) {
      const generator = new Generator(supertype, options);
      generator.generate();
    }

    return new Entrypoint(options, new Runtime(supertype, options));
  };

  private constructor(
    private opts: CompleteEntrypointOptions,
    private runtime: Runtime<T, M>
  ) {}

  findAll(matcher: (meta: M) => boolean): T[] {
    return this.runtime.entries
      .filter(entry => matcher(entry.meta))
      .map(entry => this.runtime.load(entry).entry);
  }

  find(matcher: (meta: M) => boolean): T | undefined {
    const entry = this.runtime.entries.find(entry => matcher(entry.meta));
    if (entry) {
      return this.runtime.load(entry).entry;
    }
  }

  get all(): LoadableMetadata<T, M>[] {
    return this.runtime.entries.map(e => {
      return { ...e.meta, load: () => this.runtime.load(e).entry };
    });
  }

  codegen(filepath: string): TemplaterFunction<T, M> {
    return new Codegen(
      this.opts.sourceFile.addRelative(filepath, 'file'),
      this.opts,
      this.runtime
    ).template;
  }
}

interface Init {
  <T extends ConstructorLike, M extends MetadataLike>(
    supertype: T,
    opts: EntrypointOptions<T, M>
  ): Entrypoint<T, M>;
}
