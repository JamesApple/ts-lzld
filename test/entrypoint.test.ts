import { expect } from 'chai';
import { Entrypoint, GeneratedFileData } from '../src';
import { AbsolutePath } from '../src/Filepath';
import { makeInMemoryFiles } from '../src/Files';

const root = '/my-projects/project';
const filename = root + '/index.ts';
const metadata = '/my-projects/project/metadata.json';

const redPath = root + '/nested/path/to/red.ts';
const importableRedPath = root + '/nested/path/to/red';
const bluePath = root + '/different/path/to/blue.ts';

class BaseClass {
  static type: string;
}

class RedInstance extends BaseClass {
  static type = 'red';
}

class BlueInstance extends BaseClass {
  static type = 'blue';
}

describe('Entrypoint', () => {
  it('should not fail if generate is false on load', async function () {
    const [files] = makeInMemoryFiles({});

    const entry = Entrypoint.init(BaseClass, {
      __filename: filename,
      __private: { files },
      generate: false,
      match: /correct.ts$/,
      metadataFilepath: 'metadata.json',
    });
  });

  it('should fail if attempting to read metadata and it is invalid', async function () {
    const [files] = makeInMemoryFiles({
      modules: {
        [metadata]: {},
      },
    });

    const entry = Entrypoint.init(BaseClass, {
      __filename: filename,
      __private: { files },
      generate: false,
      match: /correct.ts$/,
      metadataFilepath: 'metadata.json',
    });

    expect(() => entry.find(() => true)).to.throw(
      `lzld: invalid metadata file referenced at ${metadata} from ${filename}`
    );
  });

  let generated: GeneratedFileData;
  it('should generate the metadata file', async function () {
    const [files, data] = makeInMemoryFiles({
      modules: {
        [redPath]: { redInstance: RedInstance },
        [bluePath]: { default: BlueInstance },
      },
      recurseResult: {
        [root]: [
          AbsolutePath.fromAbsolute(redPath, 'file'),
          AbsolutePath.fromAbsolute(bluePath, 'file'),
        ],
      },
    });

    Entrypoint.init(BaseClass, {
      generate: true,
      __filename: filename,
      __private: { files },
      match: /\/([a-zA-Z0-9]+).ts$/,
      metadataFilepath: 'metadata.json',
      getMetadata: (target, { result: [, name] }) => ({
        name,
        type: target.type,
      }),
    });

    expect(data.createdFiles[metadata]).to.exist;
    generated = JSON.parse(data.createdFiles[metadata]);
    expect(generated).to.matchSnapshot();
  });

  it('should succeed in loading a module', async function () {
    const [files] = makeInMemoryFiles({
      modules: {
        [importableRedPath]: { redInstance: RedInstance },
        [metadata]: generated,
      },
    });

    const entry = Entrypoint.init(BaseClass, {
      __filename: filename,
      __private: { files },
      generate: false,
      match: /correct.ts$/,
      metadataFilepath: 'metadata.json',
      getMetadata: (target, { result: [, name] }) => ({
        name,
        type: target.type,
      }),
    });

    const Instance = entry.find(({ type }) => type === 'red');
    expect(Instance).to.equal(RedInstance);

    // Missing entry
    expect(() => entry.find(({ type }) => type === 'blue')).to.throw(
      'No module for /my-projects/project/different/path/to/blue'
    );
  });
});
