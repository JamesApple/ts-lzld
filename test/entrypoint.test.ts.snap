// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entrypoint should generate the metadata file 1`] = `
Object {
  "//": "This file is auto-generated in 'index.ts'",
  "entries": Array [
    Object {
      "meta": Object {
        "name": "red",
        "type": "red",
      },
      "name": "redInstance",
      "path": "./nested/path/to/red",
    },
    Object {
      "meta": Object {
        "name": "blue",
        "type": "blue",
      },
      "name": "default",
      "path": "./different/path/to/blue",
    },
  ],
  "lzld": "generated",
}
`;
